pkg/genericcomp/genericcomp.go

```diff
@@ -0,0 +1,148 @@
+func NewService(generic v1beta1v8o.VerrazzanoGenericComponent, namespace string, labels map[string]string) *corev1.Service {
+	service := &corev1.Service{}
+
+	// Add all container ports to the k8s service
+	for _, container := range generic.Deployment.Containers {
+		for _, port := range container.Ports {
+			sp := corev1.ServicePort{
+				Name:     port.Name,
+				Protocol: port.Protocol,
+				Port:     port.ContainerPort,
+			}
+			service.Spec.Ports = append(service.Spec.Ports, sp)
+		}
+	}
+
+	// No ports then no need for a service
+	if len(service.Spec.Ports) == 0 {
+		return nil
+	}
+
+	service.Name = generic.Name
+	service.Namespace = namespace
+	service.Labels = labels
+
+	service.Spec.Selector = map[string]string{GenericComponentSelectorLabel: generic.Name}
+	service.Spec.Type = corev1.ServiceTypeClusterIP
+
+	return service
+}
+
+// GetSecrets returns the secrets required by a generic component deployment.
+func GetSecrets(deploy appsv1.Deployment) []string {
+	var secrets []string
+
+	// Capture any image pull secrets
+	for _, secret := range deploy.Spec.Template.Spec.ImagePullSecrets {
+		secrets = append(secrets, secret.Name)
+	}
+
+	// Capture any secrets referenced in container env variables
+	for _, container := range deploy.Spec.Template.Spec.Containers {
+		for _, env := range container.Env {
+			if env.ValueFrom != nil && env.ValueFrom.SecretKeyRef != nil {
+				secrets = append(secrets, env.ValueFrom.SecretKeyRef.Name)
+			}
+		}
+	}
+
+	// Capture any secrets referenced in init container env variables
+	for _, container := range deploy.Spec.Template.Spec.InitContainers {
+		for _, env := range container.Env {
+			if env.ValueFrom != nil && env.ValueFrom.SecretKeyRef != nil {
+				secrets = append(secrets, env.ValueFrom.SecretKeyRef.Name)
+			}
+		}
+	}
+
+	return secrets
+}
+
+// UpdateEnvVars adds environment variables to a generic components deployment container.
+func UpdateEnvVars(mc *types.ManagedCluster, component string, envs *[]corev1.EnvVar) {
+	if envs == nil || len(*envs) == 0 {
+		return
+	}
+
+	for _, generic := range mc.GenericComponents {
+		if generic.Name == component {
+			for _, deployment := range mc.Deployments {
+				if deployment.Name == component {
+					addContainerEnvs(envs, deployment.Spec.Template.Spec.Containers)
+					addContainerEnvs(envs, deployment.Spec.Template.Spec.InitContainers)
+					return
+				}
+			}
+		}
+	}
+}
+
+// Add environment variables to containers
+func addContainerEnvs(envs *[]corev1.EnvVar, containers []corev1.Container) {
+	for index, container := range containers {
+		for _, env := range *envs {
+			found := false
+			for i, cenv := range container.Env {
+				if cenv.Name == env.Name {
+					container.Env[i].Value = env.Value
+					found = true
+				}
+			}
+			// Only add env variable to container if the env variable does not already exist.
+			if !found {
+				container.Env = append(container.Env, env)
+			}
+		}
+		containers[index] = *container.DeepCopy()
```

Comment:
why is DeepCopy done?

Reply:
I was having trouble with just doing an assignment so I did a deep copy

Reply:
you don't have to do that if you use containers[index] and get rid of container.  
for index, _ := range ..
Since you have an inner env loop maybe using cIndex or something to make it clearer

Reply:
changed as suggested

Reply:
that is even better i forgot you don't need the _